File: src/main/java/tomeko/screenshotmessageenhancer/screenshots/ScreenshotManager.java
package tomeko.screenshotmessageenhancer.screenshots;

import ca.weblite.objc.Client;
import ca.weblite.objc.Proxy;

import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.File;
import java.util.ArrayList;
import java.util.Locale;
import javax.imageio.ImageIO; // Standard Java Image Loader

import net.minecraft.ChatFormatting;
import net.minecraft.client.Minecraft;
import net.minecraft.network.chat.Component;
import net.minecraft.util.Util;

public class ScreenshotManager {
    public static final ArrayList<File> screenshotFiles = new ArrayList<>();
    private static final Minecraft client = Minecraft.getInstance();

    public static void copyScreenshot(int pos) {
        if (pos >= screenshotFiles.size()) return;

        File file = screenshotFiles.get(pos);
        if (!file.exists()) return;

        Component message = Component.literal("Screenshot copied to clipboard!").withStyle(style -> style.withColor(ChatFormatting.GREEN));

        if (System.getProperty("os.name").toLowerCase(Locale.ROOT).contains("mac")) {
            Util.ioPool().execute(() -> {
                try {
                    Client macClient = Client.getInstance();
                    Proxy url = macClient.sendProxy("NSURL", "fileURLWithPath:", file.getPath());

                    Proxy image = macClient.sendProxy("NSImage", "alloc");
                    image.send("initWithContentsOfURL:", url);

                    Proxy array = macClient.sendProxy("NSArray", "array");
                    array = array.sendProxy("arrayByAddingObject:", image);

                    Proxy pasteboard = macClient.sendProxy("NSPasteboard", "generalPasteboard");
                    pasteboard.send("clearContents");
                    pasteboard.sendBoolean("writeObjects:", array);

                    client.execute(() -> sendChatMessage(message));
                } catch (Exception e) {
                    e.printStackTrace();
                }
            });
            return;
        }

        // Windows / Linux
        Util.ioPool().execute(() -> {
            try {
                // Safely read the image directly from the disk file
                BufferedImage image = ImageIO.read(file);
                if (image != null) {
                    ImageContent content = new ImageContent(image);
                    Toolkit.getDefaultToolkit().getSystemClipboard().setContents(content, null);
                    client.execute(() -> sendChatMessage(message));
                }
            } catch (Exception e) {
                Component errorMessage = Component.literal("Failed to read screenshot file for clipboard.").withStyle(style -> style.withColor(ChatFormatting.RED));
                client.execute(() -> sendChatMessage(errorMessage));
                e.printStackTrace();
            }
        });
    }

    public static void deleteScreenshot(int pos) {
        if (pos >= screenshotFiles.size()) return;

        Util.ioPool().execute(() -> {
            File file = screenshotFiles.get(pos);
            if (file.exists() && file.delete()) {
                Component message = Component.literal("Screenshot deleted!").withStyle(style -> style.withColor(ChatFormatting.RED));
                client.execute(() -> sendChatMessage(message));
            } else {
                Component errorMessage = Component.literal("Could not delete screenshot (File not found)").withStyle(style -> style.withColor(ChatFormatting.GOLD));
                client.execute(() -> sendChatMessage(errorMessage));
            }
        });
    }

    private static void sendChatMessage(Component message) {
        if (client.player != null) {
            //? if >= 26.2 {
            client.gui.hud.getChat().addClientSystemMessage(message);
            //?} else {
            /*client.gui.getChat().addClientSystemMessage(message);
             *///?}
        }
    }
}

File: src/main/kotlin/tomeko/screenshotmessageenhancer/screenshots/ScreenshotManager.kt
package tomeko.screenshotmessageenhancer.screenshots

//? if 1.8.9 {
/*import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiScreen
import net.minecraft.util.EnumChatFormatting
*///?} else {
import net.minecraft.ChatFormatting
import net.minecraft.client.Minecraft
import net.minecraft.network.chat.Component
//? if >= 1.21.11 {
import net.minecraft.util.Util
//?} else {
//import net.minecraft.Util
//?}
//?}
import tomeko.screenshotmessageenhancer.compat.getStyledChatMessage
import tomeko.screenshotmessageenhancer.compat.sendChatMessage
//? if 1.8.9 {
/*import tomeko.screenshotmessageenhancer.utils.threadExecutor
import java.awt.Desktop
*///?}

import java.awt.Toolkit
import java.awt.image.BufferedImage
import java.io.File
import java.util.ArrayList
import javax.imageio.ImageIO

object ScreenshotManager {
    val screenshotFiles: ArrayList<File> = ArrayList()

    private val mc: Minecraft =
    //? if = 1.8.9 {
    //Minecraft.getMinecraft()

        //?} else {
        Minecraft.getInstance()

    //?}
    fun copyScreenshot(pos: Int, showMessage: Boolean) {
        if (pos >= screenshotFiles.size) return

        val file = screenshotFiles[pos]
        if (!file.exists()) return

        //? if 1.8.9 {
        //threadExecutor.execute {
        //?} else {
        Util.ioPool().execute {
            //?}
            try {
                val image: BufferedImage? = ImageIO.read(file)

                if (image != null) {
                    val content = ImageContent(image)

                    Toolkit.getDefaultToolkit().systemClipboard.setContents(content, null)

                    //? if 1.8.9 {
                    //mc.addScheduledTask {
                    //?} else {
                    mc.execute {
                        //?}
                        sendChatMessage(
                            getStyledChatMessage(
                                "Screenshot copied to clipboard!",
                                //? if 1.8.9 {
                                //EnumChatFormatting.GREEN
                                //?} else {
                                ChatFormatting.GREEN
                                //?}
                            ),
                            showMessage
                        )
                    }
                }
            } catch (e: Exception) {
                //? if 1.8.9 {
                //mc.addScheduledTask {
                //?} else {
                mc.execute {
                    //?}
                    sendChatMessage(
                        getStyledChatMessage(
                            "Failed to read screenshot file for clipboard.",
                            //? if 1.8.9 {
                            //EnumChatFormatting.RED
                            //?} else {
                            ChatFormatting.RED
                            //?}
                        ),
                        showMessage
                    )
                }

                e.printStackTrace()
            }
        }
    }

    fun deleteScreenshot(pos: Int) {
        if (pos >= screenshotFiles.size) return

        //? if 1.8.9 {
        //threadExecutor.execute {
        //?} else {
        Util.ioPool().execute {
            //?}
            val file = screenshotFiles[pos]

            if (file.exists() && file.delete()) {
                //? if 1.8.9 {
                //mc.addScheduledTask {
                //?} else {
                mc.execute {
                    //?}
                    sendChatMessage(
                        getStyledChatMessage(
                            "Screenshot deleted!",
                            //? if 1.8.9 {
                            //EnumChatFormatting.RED
                            //?} else {
                            ChatFormatting.RED
                            //?}
                        ),
                        true
                    )
                }
            } else {
                //? if 1.8.9 {
                //mc.addScheduledTask {
                //?} else {
                mc.execute {
                    //?}
                    sendChatMessage(
                        getStyledChatMessage(
                            "Couldn't delete screenshot (File not found)",
                            //? if 1.8.9 {
                            //EnumChatFormatting.GOLD
                            //?} else {
                            ChatFormatting.GOLD
                            //?}
                        ),
                        true
                    )
                }
            }
        }
    }

    //? if 1.8.9 {
    /*fun openScreenshot(pos: Int) {
        if (pos >= screenshotFiles.size) return

        val file = screenshotFiles[pos]
        if (!file.exists()) return

        try {
            if (Desktop.isDesktopSupported() && Desktop.getDesktop().isSupported(Desktop.Action.OPEN)) {
                Desktop.getDesktop().open(file)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun openScreenshotFolder() {
        val folder = File(Minecraft.getMinecraft().mcDataDir, "screenshots")

        if (!folder.exists()) {
            folder.mkdirs()
        }

        try {
            if (Desktop.isDesktopSupported() && Desktop.getDesktop().isSupported(Desktop.Action.OPEN)) {
                Desktop.getDesktop().open(folder)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
    *///?}

    fun uploadScreenshot(pos: Int) {
        if (pos >= screenshotFiles.size) return

        val file = screenshotFiles[pos]
        if (!file.exists()) return

        sendChatMessage(
            getStyledChatMessage(
                "Uploading screenshot...",
                //? if 1.8.9 {
                //EnumChatFormatting.YELLOW
                //?} else {
                ChatFormatting.YELLOW
                //?}
            ),
            true
        )

        //? if 1.8.9 {
        //threadExecutor.execute {
        //?} else {
        Util.ioPool().execute {
            //?}
            try {
                ScreenshotUploader.upload(file).thenAccept { url ->
                    //? if 1.8.9 {
                    //GuiScreen.setClipboardString(url)
                    //?} else {
                    mc.keyboardHandler.clipboard = url
                    //?}

                    //? if 1.8.9 {
                    //mc.addScheduledTask {
                    //?} else {
                    mc.execute {
                        //?}
                        sendChatMessage(
                            getStyledChatMessage(
                                "Screenshot uploaded + link copied to clipboard!",
                                //? if 1.8.9 {
                                //EnumChatFormatting.YELLOW
                                //?} else {
                                ChatFormatting.YELLOW
                                //?}
                            ),
                            true
                        )
                    }
                }.exceptionally { error ->
                    error.printStackTrace()
                    null
                }
            } catch (_: Exception) {
                //? if 1.8.9 {
                //mc.addScheduledTask {
                //?} else {
                mc.execute {
                    //?}
                    sendChatMessage(
                        getStyledChatMessage(
                            "Failed to upload screenshot!",
                            //? if 1.8.9 {
                            //EnumChatFormatting.RED
                            //?} else {
                            ChatFormatting.RED
                            //?}
                        ),
                        true
                    )
                }
            }
        }
    }
}